src/app/EventManagement.cpp

```diff
@@ -0,0 +1,860 @@
+/**
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2015-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/CriticalSection.h>
+#include <app/EventManagement.h>
+#include <app/InteractionModelEngine.h>
+#include <core/CHIPEventLoggingConfig.h>
+#include <core/CHIPTLVUtilities.hpp>
+#include <inttypes.h>
+#include <support/CodeUtils.h>
+#include <support/ErrorStr.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemTimer.h>
+
+using namespace chip::TLV;
+
+namespace chip {
+namespace app {
+static EventManagement sInstance;
+
+/**
+ * @brief
+ *   A TLVReader backed by CircularEventBuffer
+ */
+class CircularEventReader : public TLV::TLVReader
+{
+public:
+    /**
+     * @brief
+     *   Initializes a TLVReader object backed by CircularEventBuffer
+     *
+     * Reading begins in the CircularTLVBuffer belonging to this
+     * CircularEventBuffer.  When the reader runs out of data, it begins
+     * to read from the previous CircularEventBuffer.
+     *
+     * @param[in] apBuf A pointer to a fully initialized CircularEventBuffer
+     *
+     */
+    void Init(CircularEventBufferWrapper * apBuf);
+
+    virtual ~CircularEventReader() = default;
+};
+
+EventManagement & EventManagement::GetInstance(void)
+{
+    return sInstance;
+}
+
+struct ReclaimEventCtx
+{
+    CircularEventBuffer * mpEventBuffer = nullptr;
+    size_t mSpaceNeededForMovedEvent    = 0;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing event list.
+ */
+struct CopyAndAdjustDeltaTimeContext
+{
+    CopyAndAdjustDeltaTimeContext(TLVWriter * aWriter, EventLoadOutContext * inContext) : mpWriter(aWriter), mpContext(inContext) {}
+
+    TLV::TLVWriter * mpWriter       = nullptr;
+    EventLoadOutContext * mpContext = nullptr;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing events.
+ */
+struct EventEnvelopeContext
+{
+    EventEnvelopeContext() {}
+
+    uint16_t mFieldsToRead = 0;
+    /* PriorityLevel and DeltaSystemTimestamp are there if that is not first event when putting events in report*/
+    Timestamp mDeltaSystemTime = Timestamp::System(0);
+    Timestamp mDeltaUtc        = Timestamp::UTC(0);
```

Comment:
Is there a reason that we have the mDeltaUtc field?  It appears to be unused.

Reply:
from spec, it has utc and system timestamp and delta utc and detla system timestamp, currently i only implement system time, there is one TODO for delta UTC support